apps/server-asset-sg/src/utils/process-queue.ts

```diff
@@ -0,0 +1,82 @@
+import { sleep } from '@asset-sg/shared/v2';
+
+/**
+ * `ProcessQueue` is an async task runner that limits the number of tasks that can be run in parallel.
+ */
+export class ProcessQueue {
+  private readonly buffer: Process[] = [];
+  private readonly active = new Map<Process, Promise<void>>();
+
+  constructor(private readonly quantity: number) {
+    if (this.quantity < 1) {
+      throw new Error('Quantity must a positive integer.');
+    }
+  }
+
+  /**
+   * Registers a new tasks to be run.
+   *
+   * @param run The task to run.
+   */
+  add(run: () => void | Promise<void>): Promise<void> {
+    if (this.active.size < this.quantity) {
+      return new Promise((resolve) => {
+        const process: Process = { run, resolve };
+        this.run(process).then();
```

Comment:
We do not handle rejections explicitly, so the process would remove itself and any exception would be logged to the error stream, right?